File: Kotlin/ch08/src/main/kotlin/ch08/SimpleHashTable.kt
package ch08

/**
 * 단순 해시 테이블 (Simple Hash Table) — 직접 주소화
 * Rust/ch08/src/simple_hash_table.rs → Kotlin 이디엄 포팅
 */
class SimpleHashTable(val tableSize: Int) {
    private val keys = IntArray(tableSize)
    private val values = IntArray(tableSize)

    fun hash(key: Int): Int = key % tableSize

    fun set(key: Int, value: Int) {
        val addr = hash(key)
        keys[addr] = key
        values[addr] = value
    }

    fun get(key: Int): Int = values[hash(key)]
}

fun main() {
    val ht = SimpleHashTable(193)
    ht.set(418, 32114); ht.set(9, 514); ht.set(27, 8917); ht.set(1031, 286)
    println(ht.get(418)); println(ht.get(9)); println(ht.get(27)); println(ht.get(1031))
}

File: Rust/ch08/src/simple_hash_table.rs
// 단순 해시 테이블 (Simple Hash Table)
// Clang/08/SimpleHashTable 포팅. 직접 주소화이므로 Vec 사용.
pub type KeyType = i32;
pub type ValueType = i32;

#[derive(Clone, Copy, Default)]
pub struct Node {
    pub key: KeyType,
    pub value: ValueType,
}

pub struct HashTable {
    pub table_size: i32,
    pub table: Vec<Node>,
}

impl HashTable {
    pub fn create(table_size: i32) -> HashTable {
        HashTable {
            table_size,
            table: vec![Node::default(); table_size as usize],
        }
    }

    pub fn hash(key: KeyType, table_size: i32) -> i32 {
        key % table_size
    }

    pub fn set(&mut self, key: KeyType, value: ValueType) {
        let address = HashTable::hash(key, self.table_size);
        self.table[address as usize].key = key;
        self.table[address as usize].value = value;
    }

    pub fn get(&self, key: KeyType) -> ValueType {
        let address = HashTable::hash(key, self.table_size);
        self.table[address as usize].value
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn set_then_get_returns_value() {
        let mut ht = HashTable::create(193);
        ht.set(418, 32114);
        ht.set(9, 514);
        ht.set(27, 8917);
        ht.set(1031, 286);
        assert_eq!(ht.get(418), 32114);
        assert_eq!(ht.get(9), 514);
        assert_eq!(ht.get(27), 8917);
        assert_eq!(ht.get(1031), 286);
    }

    #[test]
    fn hash_is_key_mod_size() {
        assert_eq!(HashTable::hash(418, 193), 418 % 193);
        assert_eq!(HashTable::hash(9, 193), 9);
    }
}
